### PR TITLE
[JENKINS-31575] Adding SSO/Client certificate authentication against Active Directory

### PR DESCRIPTION
**Pull request by llecaroz**
bind dn username/password are mandatory when working with sso except in
no_check mode. In the case of no_check mode:
-A suffix will be added on username not to check against LDAP & mix them with LDAP real users authentications. 
-Users logged in SSO will be part of the default group defined in the configuration.
For all other modes, user groups will be those downloaded from LDAP. Because depending on companies, SSO DN username can differ from LDAP username (lower/upper case issues), a strategy can be applied.

For more information, see new options in the Jenkins Global security configuration

### CODE REVIEW

**Comment by llecaroz** [comment]
This pull request also contains some fixs on some findjobs warnings/errors coming from the master ci-connect branch of this projects (this was really annoying of having red errors and no checks correctly passed)

**Comment by llecaroz** [comment]
Jira ticket created as recommended : [JENKINS-31575 Adding SSO/Client certificate authentication against Active Director](https://issues.jenkins-ci.org/browse/JENKINS-31575)

**Comment by batmat** [comment]
Interesting work. Thanks.

Some quick remarks (I'm not the maintainer, just saw your mail on the list and gave it a quick shot).

/me think you should rewrite commits so that they have an user.name. Would be better for future potential forensics. And potentially squash your multiple merge commits so that the PR is more logically separated.

And it's generally common practice to try and separate PRs between features and (apparently) unrelated change (here Findbugs). This makes a bigger chance for your PR to be accepted, since it makes it easier to review.

My 2 cents

File: src/main/java/hudson/plugins/active_directory/AbstractActiveDirectoryAuthenticationProvider.java [hunk 1]
```diff
@@ -45,13 +53,48 @@ protected AbstractActiveDirectoryAuthenticationProvider() {
...
     protected abstract boolean canRetrieveUserByName();
 
+    public GroupDetails loadGroupByGroupname(String groupname) {
+      if(
+            this.realm!=null && 
+            this.realm.getSsoEnabled()!=null && 
+            this.realm.getSsoEnabled().getCheckMode()!=null && 
+            this.realm.getSsoEnabled().getCheckMode().equalsIgnoreCase("no_check")==true && 
+            this.realm.getSsoEnabled().getDefaultGroup()!=null  && 
+            this.realm.getSsoEnabled().getDefaultGroup().equals(groupname)==true)
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/AbstractActiveDirectoryAuthenticationProvider.java]: a matter of preference maybe, but when conditions get complex, I generally extract it to a small dedicated boolean method. Something like isBlahMode() or hasSomethingSpecificBluhEnabled().

File: src/main/java/hudson/plugins/active_directory/AbstractActiveDirectoryAuthenticationProvider.java [hunk 2]
```diff
@@ -45,13 +53,48 @@ protected AbstractActiveDirectoryAuthenticationProvider() {
...
+    }
+   
     public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException, DataAccessException {
-        return retrieveUser(username,null);
+      // SSO Enabled in no_check mode, find suffix and return ActiveDirectoryDetails
+      if(
+            this.realm!=null && 
+            this.realm.getSsoEnabled()!=null && 
+            this.realm.getSsoEnabled().getCheckMode()!=null && 
+            this.realm.getSsoEnabled().getCheckMode().equalsIgnoreCase("no_check")==true) {
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/AbstractActiveDirectoryAuthenticationProvider.java]: == true ?

File: src/main/java/hudson/plugins/active_directory/AbstractActiveDirectoryAuthenticationProvider.java [hunk 3]
```diff
@@ -45,13 +53,48 @@ protected AbstractActiveDirectoryAuthenticationProvider() {
...
      * Returns true if we can retrieve user just from the name without supplying any credential.
      */
     protected abstract boolean canRetrieveUserByName();
 
+    public GroupDetails loadGroupByGroupname(String groupname) {
+      if(
+            this.realm!=null && 
+            this.realm.getSsoEnabled()!=null && 
+            this.realm.getSsoEnabled().getCheckMode()!=null && 
+            this.realm.getSsoEnabled().getCheckMode().equalsIgnoreCase("no_check")==true &&
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/AbstractActiveDirectoryAuthenticationProvider.java]: Idem.

And wouldn't it be shorter using the typical pattern: constant.equals(var) instead of var.equals(constant).

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 4]
```diff
@@ -92,7 +106,54 @@
  * @author Kohsuke Kawaguchi
  */
 public class ActiveDirectorySecurityRealm extends AbstractPasswordBasedSecurityRealm {
+  static public class SSOOptions {
+    public final String userField;
+    public final String defaultGroup;
+    public final String checkMode;
+    public final String jenkinsUsernameSuffix;
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: all public? Didn't you put getters below?

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 5]
```diff
@@ -92,7 +106,54 @@
...
+    
+    public String getJenkinsUsernameSuffix()
+    {
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: wrong location for curly brace

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 6]
```diff
@@ -92,7 +106,54 @@
...
+      this.userField=userField;
+      this.checkMode=checkMode;
+      this.jenkinsUsernameSuffix=jenkinsUsernameSuffix;
+      this.defaultGroup=defaultGroup;
+    }
+  }
+  
+  public final SSOOptions ssoEnabled;
+  /**
+   * Is SSO Enabled
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: IMO this is not readable enough. Maybe that code started with only a boolean and got refactored? I don't see why this variable should be called "ssoEnabled"

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 7]
```diff
@@ -182,6 +245,62 @@ public GroupLookupStrategy getGroupLookupStrategy() {
...
+                  request.getAttribute("javax.servlet.request.X509Certificate");
+                
+                final Authentication a;
+                if (certChain != null && certChain[0] != null && ActiveDirectorySecurityRealm.this.getSsoEnabled()!=null) {
+                    final String dn = certChain[0].getSubjectDN().getName();
+
+                    if(ActiveDirectorySecurityRealm.this.getSsoEnabled().getUserField()!=null) {
+                      final String[] splittedDn=dn.split(ActiveDirectorySecurityRealm.this.getSsoEnabled().getUserField() + "=");
+                      String username = splittedDn.length>=2?splittedDn[1].split(",")[0]:null;
+                      if(username!=null && username.isEmpty()==false) {
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: ==false unnecessary

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 8]
```diff
@@ -208,8 +327,11 @@ public Authentication autoLogin(HttpServletRequest request, HttpServletResponse
                 }
             }
         };
-        rms.setUserDetailsService(uds);
-        rms.setKey(Hudson.getInstance().getSecretKey());
+        rms.setUserDetailsService(uds);        
+        {
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: wrong formating?

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 9]
```diff
@@ -668,7 +790,7 @@ public AbstractActiveDirectoryAuthenticationProvider getAuthenticationProvider()
     @Override
     public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException, DataAccessException {
         // delegate to one of our ActiveDirectory(Unix)?AuthenticationProvider
-        return getAuthenticationProvider().loadUserByUsername(username);
+      return getAuthenticationProvider().loadUserByUsername(username);
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: unnecessary change IIUC

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 10]
```diff
@@ -655,7 +777,7 @@ public int compareTo(PrioritizedSocketInfo that) {
 
     @Override
     public GroupDetails loadGroupByGroupname(String groupname) throws UsernameNotFoundException, DataAccessException {
-        return getAuthenticationProvider().loadGroupByGroupname(groupname);
+      return getAuthenticationProvider().loadGroupByGroupname(groupname);
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: unnecessary change IIUC

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUserDetail.java [hunk 11]
```diff
@@ -117,27 +115,34 @@ public int hashCode() {
...
-                        String groupName = group.getAuthority();
-                        if (groupName != null && referencedGroups.contains(groupName.toLowerCase())) {
-                            relevantGroups.add(group);
-                        }
-                    }
-                    authorities = relevantGroups.toArray(new GrantedAuthority[relevantGroups.size()]);
-                }
-            }
+        final Jenkins instance = Jenkins.getInstance();
+        if(instance!=null) {
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUserDetail.java]: Depending on what to expect, you may be interested in using Jenkins.getActiveInstance() instead to remove the null check above

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUserDetail.java [hunk 12]
```diff
@@ -117,27 +115,34 @@ public int hashCode() {
...
+          final SecurityRealm realm = instance.getSecurityRealm();
+          if ((realm instanceof ActiveDirectorySecurityRealm)) {
+              final ActiveDirectorySecurityRealm activeDirectoryRealm = (ActiveDirectorySecurityRealm)realm;
+              if (activeDirectoryRealm.removeIrrelevantGroups) {
+                  Set<String> referencedGroups = new HashSet<String>();
+                  {
```
Comment (batmat) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUserDetail.java]: local block? because you want to make authorizationStrategy & group local to this block or something?

File: src/main/resources/hudson/plugins/active_directory/ActiveDirectorySecurityRealm/configAdvanced.jelly [hunk 13]
```diff
@@ -19,6 +19,59 @@
       <f:checkbox />
     </f:entry>
   </f:advanced>
+  <f:optionalBlock title="${%SSO Enabled}" field="ssoEnabled" name="ssoEnabled"  checked="${instance.ssoEnabled!=null}">
+    <f:entry title="Portion of DN to use for the username">
```
Comment (batmat) [inline src/main/resources/hudson/plugins/active_directory/ActiveDirectorySecurityRealm/configAdvanced.jelly]: Shouldn't this title also be localizable?

**Comment by llecaroz** [comment]
Hi batmat, I took care of most of your feedbacks & integrated in my last changes. I really appreciated your feedbacks & fully agree with most of them.
